thorlcr/thorutil/thbuf.cpp

```diff
@@ -1658,6 +1688,494 @@ ISharedSmartBuffer *createSharedSmartMemBuffer(CActivityBase *activity, unsigned
+        {
+            allocator.setown(owner.activity.getRowAllocator(owner.meta, roxiemem::RHFunique));
+        }
+        ~COutputRowStream()
+        {
+            freeRows();
+        }
+        rowcount_t queryLastKnownAvailable() const
+        {
+            return lastKnownAvailable;
+        }
+        void setLastKnownAvailable(rowcount_t _lastKnownWritten)
+        {
+            lastKnownAvailable = _lastKnownWritten;
+        }
+        void cancel()
+        {
+            eof = true;
+        }
+        void reset()
+        {
+            freeRows();
+            ds.setStream(nullptr);
+            iFileIO.clear();
+            inputStream.clear();
+            eof = false;
+            currentRow = 0;
+            lastKnownAvailable = 0;
+        }
+        virtual const void *nextRow() override
+        {
+            if (eof)
+                return nullptr;
+            else if (localRowsIndex < rows.size()) // NB: no longer used after inputStream is set
+            {
+                currentRow++;
+                return getClearRow(localRowsIndex++);
+            }
+            else if (inputStream)
+                return getRowFromStream(); // NB: will increment currentRow
+            else
+            {
+                localRowsIndex = 0;
+                rows.clear();
+
+                if (owner.getRowsInMem(rows, lastKnownAvailable))
+                {
+                    if (rows.empty())
+                    {
+                        eof = true;
+                        return nullptr;
+                    }
+                    else
+                    {
+                        currentRow++;
+                        return getClearRow(localRowsIndex++);
+                    }
+                }
+                else
+                {
+                    auto [_inputStream, _iFileIO] = owner.getReadStream();
+                    inputStream.setown(_inputStream);
+                    iFileIO.setown(_iFileIO);
+                    ds.setStream(inputStream);
+                    return getRowFromStream(); // NB: will increment currentRow
+                }
+            }
+        }
+        virtual void stop() override
+        {
+            freeRows();
+            ds.setStream(nullptr);
+
+            if (inputStream)
+            {
+                StringBuffer tracing;
+                getFileIOStats(tracing, iFileIO);
+                owner.activity.ActPrintLog("CSharedFullSpillingWriteAhead::COutputRowStream: input stream finished: output=%u%s", whichOutput, tracing.str());
+
+                iFileIO.clear();
+                inputStream.clear();
+            }
+
+            // NB: this will set lastKnownAvailable to max[(rowcount_t)-1] (within owner.readAheadCS) to prevent it being considered as lowest any longer
+            owner.outputStopped(whichOutput);
+
+            eof = true;
+        }
+    };
+    class COutputStreamSerializer : public CSimpleInterfaceOf<IRowSerializerTarget>
```

Comment:
Can you move this to same place as CThorStreamDeserializerSource - so it can be reused in the unit tests?